include/ttmlir/AffineMapUtils.h

```diff
@@ -364,6 +368,1373 @@ buildPhysicalToDeviceMap(mlir::ArrayRef<int64_t> physicalShape,
+  case mlir::AffineExprKind::FloorDiv:
+  case mlir::AffineExprKind::CeilDiv: {
+    auto rhsConst = llvm::dyn_cast<mlir::AffineConstantExpr>(rhs);
+    if (!rhsConst || rhsConst.getValue() <= 0) {
+      return 1; // Cannot analyze, conservative fallback
+    }
+    int64_t divisor = rhsConst.getValue();
+
+    // Extract multiplier and offset from LHS: looking for M*d + C pattern
+    int64_t multiplier = 0;
+    bool foundDim = false;
+
+    // Helper to check if steps exceed dim bounds
+    auto checkBoundsAndReturn = [&](int64_t stepsNeeded) -> int64_t {
+      int64_t targetDimBound = dimBounds.lookup(dimPos);
+      if (stepsNeeded >= targetDimBound) {
+        return -1; // Unconstrained - always contiguous within bounds
+      }
+      return stepsNeeded;
+    };
+
+    // Pattern 1: d_dimPos floorDiv N -> need N steps to change output
+    if (auto dimExpr = llvm::dyn_cast<mlir::AffineDimExpr>(lhs)) {
+      if (dimExpr.getPosition() == dimPos) {
+        return checkBoundsAndReturn(divisor);
+      } else {
+        // unrelated dim by itself poses no constraints
+        return -1;
+      }
+    }
+
+    if (auto lhsMod = mlir::dyn_cast<mlir::AffineBinaryOpExpr>(lhs);
+        lhsMod && lhsMod.getKind() == mlir::AffineExprKind::Mod) {
+
+      // strip away mod, it doesn't impact analysis
+      mlir::AffineExpr fusedExpr;
+      if (binOp.getKind() == mlir::AffineExprKind::FloorDiv) {
+        fusedExpr = lhsMod.getLHS().floorDiv(divisor);
+      } else {
+        fusedExpr = lhsMod.getLHS().ceilDiv(divisor);
+      }
+
+      return analyzeGridResultExprForDiscontinuity(fusedExpr, dimBounds,
+                                                   dimPos);
+    }
+
+    // Deal with nested floordiv expressions
+    if (auto lhsBinOp = llvm::dyn_cast<mlir::AffineBinaryOpExpr>(lhs);
+        lhsBinOp && lhsBinOp.getKind() == mlir::AffineExprKind::FloorDiv) {
+
+      auto innerRhsConst =
+          llvm::dyn_cast<mlir::AffineConstantExpr>(lhsBinOp.getRHS());
+      TT_assertv(innerRhsConst, "Nested binary op RHS must be constant");
+
+      int64_t combinedDivisor = innerRhsConst.getValue() * divisor;
+      return analyzeGridResultExprForDiscontinuity(
+          lhsBinOp.getLHS().floorDiv(combinedDivisor), dimBounds, dimPos);
+    }
+
+    // Pattern 2: (M * d_dimPos) floorDiv N -> need ceil(N/M) steps
+    if (auto maybeMul = isBinaryMul(lhs)) {
+      auto mulRhsConst =
+          llvm::dyn_cast<mlir::AffineConstantExpr>(maybeMul.value().getRHS());
+      if (!mulRhsConst) {
+        return 1; // Mul RHS must be constant, cannot analyze otherwise
+      }
+      int64_t mulValue = mulRhsConst.getValue();
+      // Now we know RHS is a constant, just check LHS for dim
+      if (exprIsSpecificDimExpr(maybeMul.value().getLHS(), dimPos)) {
+        // ceil(divisor / mulValue)
+        int64_t stepsNeeded = (divisor + mulValue - 1) / mulValue;
+        return checkBoundsAndReturn(stepsNeeded);
+      } else {
+        return -1;
+      }
```

Comment:
:warning: **llvm\-else\-after\-return** :warning:
do not use `` else `` after `` return ``
```suggestion
      }         return -1;
     
```